### ISSUE TITLE
VIP: allow generic decimal types

### ISSUE THREAD

**Issue by charles-cooper** [vyperlang/vyper#1807]
## Simple Summary
Something like `fixed(168, 10)` after the ABI spec.

## Motivation
TODO

## Specification
TODO

## Backwards Compatibility
TODO

## Dependencies
TODO

## Copyright
Copyright and related rights waived via [CC0](https://creativecommons.org/publicdomain/zero/1.0/)

**Comment by fubuloubu** [comment]
I don't think that you should have to specify the amount of bits the type takes up. We can calculate that.

**Comment by fubuloubu** [comment]
It would be really nice if we have the ability to create custom types (or type aliases) such that we could allow typing to be used by end users more directly, e.g. `ray := decimal(places=27)` for Maker codebase.

**Comment by fubuloubu** [comment]
Note: 38 decimal places is the most possible that can fit in 128 bits (corresponding to `fixed255x38`)

**Comment by fubuloubu** [comment]
List of decimals:
0. `fixed128x0`
1. `fixed128x0`
1. `fixed128x0`
1. `fixed128x0`
1. `fixed128x0`
1. `fixed128x0`
1. `fixed128x0`
1. `fixed128x0`
1. `fixed128x0`
1. `fixed128x0`

**Comment by fubuloubu** [comment]
Meeting notes: Pamp it

**Comment by fubuloubu** [comment]
Was thinking about this some more today. It could be something like the following:
1. Internally, Vyper users only see the `Decimal` type, which is a fixed point type that is allowed to be any width decimal (using `fixed<M>x<N>` notation) as long as the calculations align at "endpoints" (internal type conversions or ABI boundaries)
2. For interface definitions, `Decimal` takes an argument (e.g. `Decimal(fixed168x10)`) which defines the "required" shape of the `Decimal` type at that interface and pins that boundary to that particular ABI type
3. When producing ABIs, Vyper statically determines the ABI type of `Decimal` and uses `fixed<M>x<N>` notation to denote that interface type. This can be overriden using the same trick for interface definitions e.g. `def myCall(val: Decimal(fixed168x10))` to pin to a particular type (useful for ERCs)
4. All else is "undefined behavior", allowing the compiler to optimize towards the highest precision and lowest loss calculations based on all the mathematical interactions that are possible within an entire contract, given the constraints at the boundary points. This optimization should be statically determinable, so that it will always produce the same outcome given the same program (although different programs and constraint system can produce different results)

This design would be strictly better than doing uint256 math all over, with implicit assumptions about precision loss, and occasionally watching contract call failures from using SafeMath without due consideration to these scenarios. The compiler can also produce suggestions for the min/max size particular values can be at conversion boundary points in order to ensure correct operation of the algorithm.

**Comment by charles-cooper** [comment]
meeting notes: add an example, bring up for discussion again
